### PR TITLE
DEPR: DataFrame.swapaxes

### LINKED ISSUE #51946: DEPR: DataFrame.swapaxes
This is listed in the general deprecation issue https://github.com/pandas-dev/pandas/issues/18262. 

Now that we have a max dimension of 2, i.e. DataFrame (while before you could have a 3D Panel or NDFrame), I don't think there is any use case for `swapaxes` where it's not exactly the same as `transpose()`. 
It's still implemented independently from `transpose()`, but I _think_ that `df.swapaxes(0, 1)` and `df.swapaxes(1, 0)` are the only two valid options and that gives exactly the same as a `transpose()` (you can do eg `df.swapaxes(0, 0)` which is a no-op, but this generalization doesn't really seem worth keeping it as a separate method, I would say)

So let's deprecate it? (in favor of `transpose()`)

### PR DESCRIPTION
**Pull request by luke396**
- [x] closes #51946  (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/copy_view/test_methods.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def test_copy_shallow(using_copy_on_write):
         lambda df, copy: df.rename_axis(columns="test", copy=copy),
         lambda df, copy: df.astype({"b": "int64"}, copy=copy),
         # lambda df, copy: df.swaplevel(0, 0, copy=copy),
-        lambda df, copy: df.swapaxes(0, 0, copy=copy),
+        # lambda df, copy: df.swapaxes(0, 0, copy=copy),
```
Comment (luke396) [inline pandas/tests/copy_view/test_methods.py:75]: not sure some like this, should be comment or deleted

File: pandas/tests/frame/methods/test_swapaxes.py [hunk 2]
```diff
@@ -8,22 +8,30 @@
...
     def test_swapaxes_invalid_axis(self):
         df = DataFrame(np.random.randn(10, 5))
-        msg = "No axis named 2 for object type DataFrame"
-        with pytest.raises(ValueError, match=msg):
-            df.swapaxes(2, 5)
+        msg = "'DataFrame.swapaxes' is deprecated"
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            msg = "No axis named 2 for object type DataFrame"
+            with pytest.raises(ValueError, match=msg):
+                df.swapaxes(2, 5)
```
Comment (luke396) [inline pandas/tests/frame/methods/test_swapaxes.py:28]: Not sure if nested blocks of code are better expressed. (`pytest` may break pre-commit)

File: pandas/tests/copy_view/test_methods.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/copy_view/test_methods.py:75]: This should be kept

File: pandas/tests/generic/test_finalize.py [hunk 3]
```diff
@@ -251,7 +251,6 @@
             operator.methodcaller("isin", pd.DataFrame({"A": [1]})),
         ),
     ),
-    (pd.DataFrame, frame_data, operator.methodcaller("swapaxes", 0, 1)),
```
Comment (luke396) [inline pandas/tests/generic/test_finalize.py:254]: It seems that this file is testing all existing finalize methods, so I removed `swapaxes` due to its depreciation.

File: pandas/core/generic.py [hunk 4]
```diff
@@ -741,6 +741,15 @@ def swapaxes(self, axis1: Axis, axis2: Axis, copy: bool_t | None = None) -> Self
         -------
         same as input
         """
+        warnings.warn(
+            # GH#51946
+            f"'{type(self).__name__}.swapaxes' is deprecated and "
+            "will be removed in a future version. "
+            "Please use '{type(self).__name__}.transpose' instead.",
```
Comment (mroeschke) [inline pandas/core/generic.py:748]:
```suggestion
            f"Please use '{type(self).__name__}.transpose' instead.",
```

File: pandas/tests/copy_view/test_methods.py [hunk 5]
```diff
@@ -115,7 +117,13 @@ def test_methods_copy_keyword(
         index = date_range("2012-01-01", freq="D", periods=3, tz="Europe/Brussels")
 
     df = DataFrame({"a": [1, 2, 3], "b": [4, 5, 6], "c": [0.1, 0.2, 0.3]}, index=index)
-    df2 = method(df, copy=copy)
+
+    if "swapaxes" in inspect.getsource(method):
```
Comment (mroeschke) [inline pandas/tests/copy_view/test_methods.py:121]: Can you do `"swapaxes" in request.node.callspec.id`  instead?

File: pandas/tests/generic/test_generic.py [hunk 6]
```diff
@@ -230,8 +230,11 @@ def test_size_compat(self, frame_or_series):
     def test_split_compat(self, frame_or_series):
         # xref GH8846
         o = construct(frame_or_series, shape=10)
-        assert len(np.array_split(o, 5)) == 5
-        assert len(np.array_split(o, 2)) == 2
+        with tm.assert_produces_warning(
+            FutureWarning, match=".swapaxes' is deprecated", check_stacklevel=False
+        ):
+            assert len(np.array_split(o, 5)) == 5
```
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/tests/generic/test_generic.py:236]: Will `np.array_split` cease to work once `swapaxes` is removed?

[the remark this answers is not in the document]
Comment (luke396) [reply pandas/tests/generic/test_generic.py:236]: yes. If remove `swapaxes`, this test will work find, without the depreciation waring.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @luke396

**Pull request merged**